bios/stage-4/src/main.rs

```diff
@@ -56,13 +56,21 @@ pub extern "C" fn _start(info: &mut BiosInfo) -> ! {
         PhysAddr::new(info.kernel.start)
     };
     let kernel_size = info.kernel.len;
-    let mut frame_allocator = {
+    let mut frame_allocator = if info.ramdisk.start == 0 {
         let kernel_end = PhysFrame::containing_address(kernel_start + kernel_size - 1u64);
         let next_free = kernel_end + 1;
         LegacyFrameAllocator::new_starting_at(
             next_free,
             memory_map.iter().copied().map(MemoryRegion),
         )
+    } else {
+        let ramdisk_end =
+            PhysFrame::containing_address(PhysAddr::new(info.ramdisk.start + info.ramdisk.len));
```

Comment:
I think we should subtract 1 here to get an inclusive bound.

```suggestion
            PhysFrame::containing_address(PhysAddr::new(info.ramdisk.start + info.ramdisk.len - 1));
```